src/orion/core/worker/consumer.py

```diff
@@ -57,27 +65,90 @@ def __init__(self, experiment):
+              parameter space definition does not correspond to the version
+              of the actual user's code to be executed. Treatment is left to
+              the user. However **a fail fast solution could be detected**,
+              because **argparse** (for instance) exits with code 2,
+              when such a case happens. This argparse specific treatment
+              is disputable. In favour of this handling is that this practice
+              is traditional, but still, not enforced.
+
+           .. seealso::
+
+              Method `Consumer.interact_with_script`
+                 Code which would exit with 2, if user's script exited with 2.
+
+              `GNU bash manual, Exit Status <Exit Status>`_
+                 Quoting: *... return an exit status of 2 to indicate
+                 incorrect usage, generally invalid options or missing arguments.*
+
+        .. _Exit Status:
+           https://www.gnu.org/software/bash/manual/html_node/Exit-Status.html
+
         """
-        log.debug("### Create new temporary directory at '%s':", self.tmp_dir)
-        with tempfile.TemporaryDirectory(prefix=self.experiment.name + '_',
-                                         dir=self.tmp_dir) as workdirname:
-            log.debug("## New temp consumer context: %s", workdirname)
-            completed_trial = self._consume(trial, workdirname)
-
-        if completed_trial is not None:
-            log.debug("### Register successfully evaluated %s.", completed_trial)
-            self.experiment.push_completed_trial(completed_trial)
-        else:
-            log.debug("### Save %s as broken.", trial)
-            trial.status = 'broken'
-            Database().write('trials', trial.to_dict(),
-                             query={'_id': trial.id})
+        try:
+            log.debug("### Create new temporary directory at '%s':", self.tmp_dir)
+            self.latest_results = list()
+            returncode = None
+
+            with tempfile.TemporaryDirectory(prefix=self.experiment.name + '_',
+                                             dir=self.tmp_dir) as workdirname:
+                log.debug("## New temp consumer context: %s", workdirname)
+                returncode = self._consume(trial, workdirname)
+
+        except KeyboardInterrupt:
+            new_status = 'interrupted'
+            raise
+
+        except SystemExit:
+            new_status = 'broken'
+            Database().write('experiments',
+                             {'status': 'broken'},
+                             {'_id': self.experiment._id})
+            raise
+
+        finally:
+            trial.results = self.latest_results
+
+            if returncode == 0 or (returncode is None and trial.results):
+                log.debug("### Register successfully evaluated %s.", trial)
+                self.experiment.push_completed_trial(trial)
+            elif returncode is not None:
+                update_trial_with_status(trial, 'broken')
```

Comment:
Yes, I agree. We should remove `push_completed_trial` in `Experiment` in favor of `finalize_trial(trial, status)`. The new method would also complete the tuple (start/resume time, finish time) and push it as well, according to what we have discussed.

Reply:
`finalize_trial` or `update_trial`? The former sounds like the trial is not expected to change afterwards. This would be confusing: `finalize_trial(trial, 'suspended')`